### PR TITLE
I18n internationalization

### LINKED ISSUE #205: Enable i18n
Add a mechanism for internationalization (i18n) maybe with [vue-i18n](url), which was used in a Wegue based project solution: https://github.com/Geolicious/wegue/commit/c4af28b6fc5c97dff4baf2bae6abb30d11197cec

See also:  https://localazy.com/blog/how-to-localize-vuejs-app-with-vue-i18n-and-localazy

### PR DESCRIPTION
**Pull request by fschmenger**
This branch implements multi language support for Wegue and closes #205.

https://user-images.githubusercontent.com/46027124/127163078-cd8f11c7-1f25-4991-9abf-d959cd402007.mp4

Key features:
* 2 fold support of language packs: Wegue core language packages are located in src/locales and provide locale messages for core components. These are deep merged with app specific language packages located in 'app/locales' (templated from app-starter/locales), such that messages for app specific components can be easily added and Wegue core messages can be overridden. Per default English and German language packs have been provided.
* Languages supported by an application can be configured in app-conf.json, e.g.:
` "lang": {
    "supported": {
      "en": "English",
      "de": "Deutsch"
    },
    "fallback": "en"
  }`
* A new module 'wgu-localeswitcher' was added to change the active language at runtime.
` "wgu-localeswitcher": {
      "target": "toolbar",
      "darkLayout": true
    }`
* Language detection has been implemented on application start with the following precedence rules:
1. 'lang' query-parameter
2. Preferred browser language
3. One of the accepted browser languages
4. The fallback language declared in app-conf.json
5. English as a static fallback
* Vuetify language packs for vuetify base components have been included and are also switched at runtime.
* Module titles are now consistently supported across all module buttons and appear as a localized tooltip.
* Several module related messages declared in app-conf.json have been moved to the language packs, using the module identifier as a key.
* Layer names - previously declared in app-conf.json - are now obtained from the language packs. Under the key 'mapLayers' an entry correponding to the 'lid' has to be created.

This is quite a verbose (and also breaking) change-set, since several text based app-conf properties are now obsolete. Because there's no easy way to automigrate those, I decided to log a bunch of warnings on application start when such properties are declared.

I update the documentation and unit tests to reflect these changes. I would kindly ask you to give it an review first - especially regarding some of the discussion points in the comment below - so I can finalize those before you merge.

### CODE REVIEW

**Comment by fschmenger** [comment]
Here are some points for discussion, so you can have a word on it before I finalize those:

* The generic toggle buttons and in particular the wgu-zoomtomaxextent button support a 'text' property, which will display a text right of the button. These options have never been documented, IMHO they look fairly ugly and all buttons now display a tooltip containing the module title anyway. Would it be ok to remove this option? Alternatively if you want to preserve the feature, I can change the code to obtain the text via I18n.
* Regarding the 'lang' configuration option: As an adhoc decision I incorporated the friendly language name in the app-conf, e.g. `"en": "English"` . These are currently only displayed in the new language picker control. Should we localize the friendly language names itself? The reasoning behind not translating those, is that a user might have trouble to indentify his language, when the active locale is e.g. japanese. On the other hand it's no satisfying to introduce new texts in app-conf, when the purpose of this change-set is to get rid of those. Proposal: I could obtain the friendly language names via language packs (and we could still decide on not translating them) !?
* The language specific geocoder provider options should be derived from the active language. This requires some sort of mapping between the supported languages and the geocoder language codes. Where do we want to put the mapping (e.g. configure in the 'lang' property or somewhere specific in 'wgu-geocoder'? I didnt study geocoder code closely enough to know the implications, so any ideas welcome here.
* Regarding localization of layers. As a first shot I decided to obtain the layer names in controls where it is actually displayed. This results in some ugly construction like `{{ $t('mapLayers.' + (layer.get('langKey') || layer.get('lid')) + '.name') }}` at different places in the code. An alternate implementation idea would be resolving locale specific layer properties at a central spot, e.g. in Map component, and then bake the language specific properties into the OL layer collection. This could be done by hooking up layer-add, layer-change and locale-change event handlers. This approach would also enable a localization of other layer properties, e.g. attribution.

What we could consider later:
* Currently all available language packs are compiled into the application at build time. These are all under 'src/locales', 'app/locales', and for vuetify 'node_modules/vuetify/es5/locale'. I don't see this as a problem, as all of these are small. On the long run we could enable some lazy loading mechanism for language packs.

File: src/util/Locale.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        {}
+      );
+  },
+
+  /**
+   * Import all Wegue core language files from 'src/locales' and optionally
+   * app specific language files from 'app/locales'. Language files will be merged,
+   * such that Wegue core messages can be overridden by app messages.
+   *
+   * @returns A container with message data. Key is the language code, value contains the messages.
```
Comment (JakobMiksch) [inline src/util/Locale.js:39]: I guess the output type is an `Obejct`

```suggestion
   * @returns  {Object} A container with message data. Key is the language code, value contains the messages.
```

File: src/util/Locale.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      return LocaleUtil.isLanguageSupported(value, appConfig);
+    });
+  },
+
+  /**
+   * Get languages supported by the application, which are declared in the application context.
+   * Fallback will be {'en':'English}'
+   *
+   * @param {Object} appConfig Global application context.
+   * @returns A container with key value pairs, while key corresponds to the language code and
```
Comment (JakobMiksch) [inline src/util/Locale.js:115]: I guess it returns an `Object` (?)

```suggestion
   * @returns {Object} A container with key value pairs, while key corresponds to the language code and
```

File: src/util/Locale.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      Object.keys(appConfig.lang.supported).length > 0;
+    return isDefined ? appConfig.lang.supported : LocaleUtil.supportedLanguageFallback;
+  },
+
+  /**
+   * Get the fallback language declared in the application context.
+   * If none such it is declared, returns 'en'.
+   *
+   * @param {Object} appConfig Global application context.
+   * @returns The language code of the fallback language.
```
Comment (JakobMiksch) [inline src/util/Locale.js:131]:
```suggestion
   * @returns {string} The language code of the fallback language.
```

File: src/util/Locale.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      appConfig.lang.fallback;
+    return isDefined ? appConfig.lang.fallback : Object.keys(LocaleUtil.supportedLanguageFallback)[0]
+  },
+
+  /**
+   * Checks whether the language code is supported by the application.
+   *
+   * @param {String} code The language code
+   * @param {Object} appConfig Global application context.
+   * @returns True if the language code is supported by the application.
```
Comment (JakobMiksch) [inline src/util/Locale.js:145]:
```suggestion
   * @returns {boolean} True if the language code is supported by the application.
```

File: src/util/Object.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+/**
+ * Object related utility methods.
+ */
+const ObjectUtil = {
+
+  /**
+   * Tests whether item is an object
+   * @param {*} item The item to test.
+   * @returns True if the item is an object.
```
Comment (JakobMiksch) [inline src/util/Object.js:9]:
```suggestion
   * @returns {boolean} True if the item is an object.
```

File: src/util/Object.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    }
+  },
+
+  /**
+   * Recursively list all property paths in an object.
+   *
+   * @param {Object} item An object to extract property paths from.
+   * @param {String} path Internal usage for recursion.
+   * @returns An array containing all paths to keys of the object.
```
Comment (JakobMiksch) [inline src/util/Object.js:40]:
```suggestion
   * @returns {Array} An array containing all paths to keys of the object.
```

**Review by JakobMiksch — COMMENTED** [review]
@fschmenger thanks for this impressive feature!! Inline some minor formatting suggestions. A next comment regarding your bullet points in the description is on the way ...

**Review by JakobMiksch — CHANGES_REQUESTED** [review]
I could not get the URL `lang` parameter to work.

For example with this link: `http://localhost:8081/#z=2.7142&c=9.4292%2C35.0779&r=0&l=osm-bg%2Cearthquakes%2CShops&lang=de`

I appended `lang=de`, but it does not seem to have effect. 
Am I missing something?

**Comment by JakobMiksch** [comment]
I tested the app with the language feature and it works great so far. Really cool :sunglasses: 


> Here are some points for discussion, so you can have a word on it before I finalize those:
> 
>     * The generic toggle buttons and in particular the wgu-zoomtomaxextent button support a 'text' property, which will display a text right of the button. These options have never been documented, IMHO they look fairly ugly and all buttons now display a tooltip containing the module title anyway. Would it be ok to remove this option? Alternatively if you want to preserve the feature, I can change the code to obtain the text via I18n.
> 

Removing the not-used display text feature is fine for me. 

>     * Regarding the 'lang' configuration option: As an adhoc decision I incorporated the friendly language name in the app-conf, e.g. `"en": "English"` . These are currently only displayed in the new language picker control. Should we localize the friendly language names itself? The reasoning behind not translating those, is that a user might have trouble to indentify his language, when the active locale is e.g. japanese. On the other hand it's no satisfying to introduce new texts in app-conf, when the purpose of this change-set is to get rid of those. Proposal: I could obtain the friendly language names via language packs (and we could still decide on not translating them) !?
> 

Good thought. I would vote for the simplest solution here. Taking the names via the language-packs sounds good to me. Realistically we will only have use for German and English within the next time. And most user's probably know how their language is called in English.

>     * The language specific geocoder provider options should be derived from the active language. This requires some sort of mapping between the supported languages and the geocoder language codes. Where do we want to put the mapping (e.g. configure in the 'lang' property or somewhere specific in 'wgu-geocoder'? I didnt study geocoder code closely enough to know the implications, so any ideas welcome here.
> 
Would be good to sync language with the GeoCoder. I am also not that familiar with GeoCoder code. Maybe @justb4  has an idea. Otherwise we could convert this point to an issue and solve it in a follow up PR.

>     * Regarding localization of layers. As a first shot I decided to obtain the layer names in controls where it is actually displayed. This results in some ugly construction like `{{ $t('mapLayers.' + (layer.get('langKey') || layer.get('lid')) + '.name') }}` at different places in the code. An alternate implementation idea would be resolving locale specific layer properties at a central spot, e.g. in Map component, and then bake the language specific properties into the OL layer collection. This could be done by hooking up layer-add, layer-change and locale-change event handlers. This approach would also enable a localization of other layer properties, e.g. attribution.

For a first shot this is great. The idea to bake the names into the layer objects  sound good as well. For me it would  be fine to solve this in a follow-up PR.

**TLDR**
Since language support is enabled for most places in Wegue, for me it would be fine to fix some the remaining ones (see above) in follow-up PRs. 

@chrismayer what do you think?

**Comment by fschmenger** [comment]
First of all, thanks for the extended review, @JakobMiksch!

Sorry for the formatting issues, in most cases this is the result of cherry-picking and merging changes over from my application repo. I usually run it through eslint with the Wegue presets for finalization. Do you use any special linting settings which catch these issues, so you don't have to go through this on every PR?

Regarding the lang query parameter: Be aware that it is only supported as a query param and not after the fragment identifier (#). So a working example would be http://localhost:8081/?lang=de#z=2.7142&c=9.4292%2C35.0779&r=0&l=osm-bg%2Cearthquakes%2CShops. This is currently the same behaviour as for the "appCtx" parameter. If this is a common source of trouble I would propose to change the support of query parameter syntax consistently, maybe as a follow up issue.

I realized the "path is undefined" problem too. I think this is not related to the changes in this PR - but I could be wrong. Anyway I already have started a maintainance branch with a few minor fixes which I will provide after this PR. The fix for this error message is already in there.

I will follow your recommendations specific to the comments below and will also try bake the layer name (and localized attribution) into the OL layer collection. 

Regarding the "lang" definition in app-conf: One option would be to have the geocoder language mapped inside the "lang" field, something along these lines - but I'm not sure if this information is sufficient:
`"lang": {
    "supported": [
      { "code": "de", longCode: "de-DE" }
    ],
    "fallback": "en"
  }`
Alternatively we could turn "supported" into a plain array and we put the mapping somewhere into geocoder, e.g.
`"lang": {
    "supported": [
      "en",
      "de"
    ],
    "fallback": "en"
  }`

Until we have finally decided on this I would keep the "lang" field as it is - with the inclusion of the friendly language name - so I wont have to change all the implementations (including unit tests and docs) twice.

File: docs/map-layer-configuration.md [hunk 7]
```diff
@@ -37,8 +36,8 @@ The following properties can be applied to all map layer types
 | selectable          | Boolean value, whether the features of the layer can be selected by click in order to display the attributes in a window | `"selectable": true` |
 | hoverable           | Boolean value, whether the features of the layer can be hovered in order to display an attribute (see `hoverAttribute`) in a tooltip  | `"hoverable": true` |
 | hoverAttribute      | Attribute to be shown if a feature of the layer is hovered. Only has an effect if `hoverable` is set to `true`  | `"hoverAttribute": "name"` |
-| style               | Object to define a rendering style for the features of the layer  | see [style](map-layer-configuration?id=style-for-vectorlayers) |
-| selectStyle         | The style for a selected feature | see [style](map-layer-configuration?id=style-for-vectorlayers) |
+| style               | Object to define a rendering style for the features of the layer  | see [style](map-layer-configuration.md?id=style-for-vectorlayers) |
```
[the remark this answers is not in the document]
Comment (fschmenger) [reply docs/map-layer-configuration.md:39]: I tried to view the documentation on github via https://github.com/meggsimum/wegue/blob/master/docs/map-layer-configuration.md. You can see that clicking the links results in 404.

**Comment by fschmenger** [comment]
I have one more question regarding the code in factory/Layer.js, line 60ff which I currently marked with a TODO:
```
// apply LID (Layer ID) if not existent
// TODO
//  Review this, as of the localization change the lid attribute must be provided
//  and the previously used name property to compose a unique lid has been deprecated.
if (!lConf.lid) {
  // Make a unique layerId from Layer URL so contexts
  // like permalinks can be reapplied.
  lConf.lid = btoa(lConf.url).substr(0, 6);
}
```

The 'lid' if missing was previously derived from 'url' and 'name' field.
Frankly I don't see how to handle this otherwise, as a name attribute is no longer present.

If ok for you, for the very least I will finalize this by removing the TODO and adding an assertion in migrateAppConfig, that 'lid' is a mandatory field (It's already stated in the docs).

Option 2 would be to remove the `!lConf.lid` block from the factory entirely. 
Not defining an 'lid' will now lead to other bugs regarding localization and proably elsewhere in the code, when the 'lid' is used as an indexer. What do you think?

**Review by JakobMiksch — APPROVED** [review]

File: docs/map-layer-configuration.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (JakobMiksch) [reply docs/map-layer-configuration.md:39]: yes that is true for viewing the pages on GitHub. But the docs are actually meant to be viewed via [docsify](https://docsify.js.org/#/) and are automatically deployed here: https://meggsimum.github.io/wegue/

The corresponding link to your example would be https://meggsimum.github.io/wegue/#/map-layer-configuration

In docsify it does not matter if the links end with `.md`. But it would be better to be consistent here. 
I would vote for not using `.md`, because it makes the url a bit cleaner.

**Comment by JakobMiksch** [comment]
> Sorry for the formatting issues, in most cases this is the result of cherry-picking and merging changes over from my application repo. I usually run it through eslint with the Wegue presets for finalization. Do you use any special linting settings which catch these issues, so you don't have to go through this on every PR?
> 

I typically use `npm run lint-fix`. It might be a good idea for the future to include the liniting into the CI-pipeline.

> Regarding the lang query parameter: Be aware that it is only supported as a query param and not after the fragment identifier (#). So a working example would be http://localhost:8081/?lang=de#z=2.7142&c=9.4292%2C35.0779&r=0&l=osm-bg%2Cearthquakes%2CShops. This is currently the same behaviour as for the "appCtx" parameter. If this is a common source of trouble I would propose to change the support of query parameter syntax consistently, maybe as a follow up issue.
> 

Thanks for clarification. I tried it and it worked like expected :+1:

File: docs/map-layer-configuration.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (fschmenger) [reply docs/map-layer-configuration.md:39]: Ok got it. I will partially revert c5b90e44d0dd3aa98255739ea1fe920db2ae748d then.

**Comment by fschmenger** [comment]
The points discussed above have been addressed and this should be ready to merge now. 

I would kindly ask you to convert the topic regarding the format of the "lang" attribute into a follow up issue (regarding friendly language name and localization of geocoder search provider).

**Review by chrismayer — COMMENTED** [review]
This is really nice. Great work @fschmenger! Important thing to have multi lang support :+1:  I had a look at the code and besides some minor cosmetic things I posted some suggestions for renaming.

Besides that one thing bothers me a bit: By now users can configure the application in one spot (app-conf.json). From the changes in this PR users need to configure in several places: app-conf, locale1, locale2, and so on. Especially that configuring map layers needs work in different files makes things IMHO complicated. Not sure how to deal with that and currently I have no final solution for that. Just to mention in other projects we offer to set texts the "normal" way (e.g. config or ViewModel) and an optional set i18n text overwrites the text. Not sure if this would be applicable here.

**Comment by chrismayer** [comment]
I also just checked out your code and tested the demo applications. Works like charm :+1:

File: src/components/attributeTable/AttributeTableWin.vue [hunk 8]
```diff
@@ -85,34 +79,24 @@ export default {
...
-            layer.get('lid') !== 'measure-layer') {
-          layerItems.push({
-            layerName: layer.get('name'),
-            lid: layer.get('lid')
-          });
-        }
-      });
-
-      this.layerItems = layerItems
+    vectorLayers () {
```
[the remark this answers is not in the document]
Comment (fschmenger) [reply src/components/attributeTable/AttributeTableWin.vue:93]: This is actually not a method but a (computed) property, thats why there is no 'get' in the name. It returns all the vector layers in the OL collection. What name would you propose instead? 'displayedLayers' maybe, or go back to 'layerItems' ?

File: src/components/layerlist/LayerList.vue [hunk 9]
```diff
@@ -13,111 +22,44 @@
...
-      layerVizChanged () {
-        var me = this;
-
-        me.layerItems.forEach(function (layerItem, idx) {
-          const layer = LayerUtil.getLayerByLid(layerItem.lid, me.map);
-          if (layer) {
-            layer.setVisible(layerItem.visible);
-          }
-        });
+      fgLayers () {
```
[the remark this answers is not in the document]
Comment (fschmenger) [reply src/components/layerlist/LayerList.vue:59]: This is actually not a method but a (computed) property, thats why there is no 'get' in the name. What would you prefer 'foregroundLayers' / 'displayedLayers' / 'layerItems' ?

File: src/components/measuretool/MeasureResult.vue [hunk 10]
```diff
@@ -54,11 +54,11 @@
         const length = getLength(line);
         let output;
         if (length > 100) {
-          output = (Math.round(length / 1000 * 100) / 100) +
-              ' ' + 'km';
+          output = this.$t('wgu-measuretool.lengthKm',
```
[the remark this answers is not in the document]
Comment (fschmenger) [reply src/components/measuretool/MeasureResult.vue:57]: It's named formatting to insert params into the language string, see https://kazupon.github.io/vue-i18n/guide/formatting.html.
The corresponding placeholder is in the language file:
```
"wgu-measuretool": {
"lengthKm": "{0} km"
}
```

Want me to add a comment in the code?

File: src/factory/Layer.js [hunk 11]
```diff
@@ -59,9 +59,9 @@ export const LayerFactory = {
   getInstance (lConf, olMap) {
     // apply LID (Layer ID) if not existent
     if (!lConf.lid) {
-      // Make a unique layerId from Layer name and URL so contexts
+      // Make a unique layerId from Layer URL so contexts
       // like permalinks can be reapplied.
-      lConf.lid = btoa(lConf.url + lConf.name).substr(0, 6);
+      lConf.lid = btoa(lConf.url).substr(0, 6);
```
[the remark this answers is not in the document]
Comment (fschmenger) [reply src/factory/Layer.js:64]: As discussed. Will be removed entirely.

**Comment by fschmenger** [comment]
@chrismayer I went through all of your suggested changes, let me know if something is still missing.

**Comment by chrismayer** [comment]
Looks good from my side. Thank you very much for this powerful feature and for you ongoing effort.

For transparency (what we discussed offline):

For layer names / layer attributions there is now no need anymore to define them in the locale files (to simplify the setup process for a custom app). Configuring the layer names / layer attributions in the app-config (as before) would be enough. In case you are configuring both for layer names / layer attributions (locale files and app-config) the values in the app-config will overwrite the ones in the locale files.

Anything to add on this @fschmenger ?

**Review by chrismayer — APPROVED** [review]

**Comment by chrismayer** [comment]
So this is good to go. Thanks again to all, who were involved.
